hosted/alibaba/helper/helper_cluster.go

```diff
@@ -0,0 +1,368 @@
+		Expect(err).To(BeNil())
+	}
+
+	if checkClusterConfig {
+		// check that the desired config is applied on Rancher
+		Eventually(func() bool {
+			ginkgo.GinkgoLogr.Info("Waiting for the node count change to appear in AliStatus.UpstreamSpec ...")
+			cluster, err = client.Management.Cluster.ByID(cluster.ID)
+			Expect(err).To(BeNil())
+			upstreamNodePools := cluster.AliStatus.UpstreamSpec.NodePools
+			for i := range upstreamNodePools {
+				if np := upstreamNodePools[i]; *np.DesiredSize != nodeCount {
+					return false
+				}
+			}
+			return true
+		}, tools.SetTimeout(helpers.Timeout), 10*time.Second).Should(BeTrue())
+	}
+
+	return cluster, nil
+}
+
+// DeleteNodePool deletes a node pool from the ALI cluster
+// if wait is set to true, it waits until the update is complete; if checkClusterConfig is true, it validates the update
+func DeleteNodePool(cluster *management.Cluster, client *rancher.Client, wait, checkClusterConfig bool) (*management.Cluster, error) {
+	currentNodePoolNumber := len(cluster.AliConfig.NodePools)
+	if len(cluster.AliConfig.NodePools) <= 1 {
+		return cluster, fmt.Errorf("cannot delete node pool: only one node pool remains")
+	}
+
+	upgradedCluster := newClusterUpdatePayload(cluster)
+	upgradedCluster.AliConfig.NodePools = upgradedCluster.AliConfig.NodePools[1:]
+
+	// Ensure ImageID is populated for all remaining node pools in the update payload
+	populateNodePoolImageIDs(upgradedCluster)
+
+	cluster, err := client.Management.Cluster.Update(cluster, upgradedCluster)
+	if err != nil {
+		return nil, err
+	}
+
+	if checkClusterConfig {
+		// Check if the desired config is set correctly
+		Expect(len(cluster.AliConfig.NodePools)).Should(BeNumerically("==", currentNodePoolNumber-1))
+		// The check for correct node pool names is complex after deletion, so we focus on the count.
+	}
+	if wait {
+		err = clusters.WaitClusterToBeUpgraded(client, cluster.ID)
+		Expect(err).To(BeNil())
+	}
+	if checkClusterConfig {
+		// Check if the desired config has been applied in Rancher
+		Eventually(func() int {
+			ginkgo.GinkgoLogr.Info("Waiting for the total node pool count to decrease in AliStatus.UpstreamSpec ...")
+			cluster, err = client.Management.Cluster.ByID(cluster.ID)
+			Expect(err).To(BeNil())
+			return len(cluster.AliStatus.UpstreamSpec.NodePools)
+		}, tools.SetTimeout(helpers.Timeout), 10*time.Second).Should(BeNumerically("==", currentNodePoolNumber-1))
+		// The check for correct node pool names is complex after deletion, so we focus on the count.
+	}
+	return cluster, nil
+}
+
+// UpgradeClusterKubernetesVersion upgrades the control plane k8s version to the value defined by upgradeToVersion.
+// If checkClusterConfig is set to true, it will validate that the control plane has been upgraded successfully
+func UpgradeClusterKubernetesVersion(cluster *management.Cluster, upgradeToVersion string, client *rancher.Client, upgradeNodePool, wait, checkClusterConfig bool) (*management.Cluster, error) {
```

Comment:
```suggestion
func UpgradeClusterKubernetesVersion(cluster *management.Cluster, upgradeToVersion string, client *rancher.Client, wait, checkClusterConfig bool) (*management.Cluster, error) {
```

`upgradeNodePool` can be removed, it is unused.